MSC4323: User suspension & locking endpoints (#4323)
matrix-org/matrix-doc: proposals/4323-agnostic-suspend-and-lock.md

Reviewer: Continuwuity has a PR that implements the suspension endpoints and authenticated capability advertisement (but not locking because that's not implemented under the hood either) - https://forgejo.ellis.link/continuwuation/continuwuity/pulls/967

provided by an entirely separate service.

This MSC will outline new endpoints that will allow server administrators to
suspend, unsuspend, lock, and unlock given users.

## Proposal

Note: what defines a "server administrator" is left up to the implementation,
most already have their own systems for defining administrators which rarely has a reason to be
exposed outside of their respective management interfaces.

Taking advantage of the 10.22 section of the client-to-server specification,
four new endpoints are introduced:

- `GET /_matrix/client/v3/admin/suspend/{userId}`
- `PUT /_matrix/client/v3/admin/suspend/{userId}`
- `GET /_matrix/client/v3/admin/lock/{userId}`
- `PUT /_matrix/client/v3/admin/lock/{userId}`

These new endpoints complement the already existing
`/_matrix/client/v3/admin/whois/{userId}` endpoint, however, are restricted to only permitting
execution on local users.

The requests and responses are outlined as below. Remember that custom properties may be
provided, given that they use proper namespaces.

The response body of both the `GET` and `PUT` endpoints, as well as the request body of the

> `PUT` endpoint should be:

Reviewer 2: I suppose there might be a mild rabbit hole on what else to add (deactivated would be an obvious third new field)

> `PUT` endpoint should be:

Reviewer 3: I'd generally favour not adding them to whois. If we instead had a general `/user/:userId` endpoint, then they'd fit there, but `whois` is intentionally narrow in scope.

> `PUT` endpoint should be:

Reviewer 3: It's something from before proposals existed, so didn't go through the same design process, making it somewhat questionable. 

It's scope is akin to a domain whois lookup though - suspension/locking/deactivation doesn't fit that description very well. Putting the fields on the whois endpoint also mean it's harder to remove/fix `/whois` later :)

**Suspend**:

```json
{
  "$schema": "https://json-schema.org/draft-07/schema",
  "type": "object",
  "properties": {
    "suspended": {"type": "boolean"}
  },
  "required": ["suspended"]
}
```

For example:

```json
{"suspended": true}
```

**Lock**:

```json
{
  "$schema": "https://json-schema.org/draft-07/schema",
  "type": "object",
  "properties": {
    "locked": {"type": "boolean"}
  },
  "required": ["locked"]
}
```

For example:

```json
{"locked": false}
```

---

Sending a `GET` to the respective endpoints should return `404` with `M_NOT_FOUND` if the
requested user ID is not found. `400` and `M_INVALID_PARAM` should be returned if the specified
user ID is valid, but not local to the server. `403` `M_FORBIDDEN` should be returned if
the user is not a server administrator.

If a user is deactivated, `M_NOT_FOUND` should also be returned, as deactivated users are not
suspended nor locked, and typically are to be permanently treated as gone.

If the server does not support the requested action, e.g. the server only supports suspension and

> not locking, the respective unsupported endpoint should return `M_NOT_FOUND` as expected.

Reviewer: I forget we have capabilities, that'd be a good idea. Advertised both that X is supported *and* can be accessed via the new endpoints

> # MSC4323: Agnostic user suspension & locking endpoints

Reviewer 4: Not really a big deal (and won't stop me from approving the MSC), but I don't understand the title. How is this agnostic? To what? Do you mean agnostic to the server implementation? That's an odd use of the word, maybe "Standardized" would make more sense.

Reviewer 4: Yep that works too!

Reviewer: Sold then, I'll remove the prefix :)

even going as far as to suggest which endpoints can and cannot be executed by suspended users.
However, it does not currently define an endpoint which server administrators can call to suspend
and unsuspend users on their server.
As a result, moderation tooling such as Draupnir and Meowlnir have to implement
implementation-specific calls, which is not sustainable as more implementations other than Synapse
integrate this feature.

This proposal will outline new endpoints that will allow server administrators to
suspend, unsuspend, lock, and unlock given users.

## Proposal

> [!IMPORTANT]
> What defines a "server administrator" is left up to the implementation itself as most already have
> their own systems for defining administrators (e.g. Synapse has a database flag, Conduit has room
> membership) which rarely has a reason to be exposed outside of their respective management
> interfaces.

Complementing [section 10.22 (Server Administration)][p1] of the client-to-server specification,
four new endpoints are introduced:

- `GET /_matrix/client/v1/admin/suspend/{userId}`
- `PUT /_matrix/client/v1/admin/suspend/{userId}`
- `GET /_matrix/client/v1/admin/lock/{userId}`

> - `PUT /_matrix/client/v1/admin/lock/{userId}`

Reviewer 5: I wonder if we should have a `/_matrix/admin` namespace instead of `/_matrix/client/v1/admin`?

I'm not sure which way makes more semantic sense, as i suppose it's accessed over the client server api?

implementation-specific versions of the suspend and lock endpoints (such as them being blocked off
at the reverse proxy) - server developers will likely want to ensure that their users are made
appropriate aware of this (via release notes or some other similar high-visibility channel) so that
they can apply those same restrictions to these new endpoints if necessary.

These endpoints also increase the compromise value of an administrator account, as an attacker who
breaches one of these accounts will be able to suspend or lock other users on the server. While
the attacker would not be able to lock out other administrators, a single-administrator homeserver
may be vulnerable to a temporary takeover if the sole administrator account is breached and
the deployment cannot be secured & have the change reverted quickly.
It may possibly also allow an attacker to disable other room-level moderation tooling, such as
moderation bots, assuming those tools are not also server administrators.

## Unstable prefix

| Stable | Unstable |
| ------ | -------- |
| `/_matrix/client/v1/admin/...` | `/_matrix/client/unstable/uk.timedout.msc4323/admin/...` |

> | `account_moderation` | `uk.timedout.msc4323` |

Reviewer 6: ```suggestion
| `m.account_moderation` | `uk.timedout.msc4323` |
```

Reviewer: I updated the highlighted area, however I don't see another place that'd need that same change?

# MSC4323: Agnostic user suspension & locking endpoints

Currently the specification outlines error codes for suspended and locked users,
even going as far as to suggest which endpoints can and cannot be executed by suspended users.
However, it does not currently define an endpoint which server administrators can call to suspend
and unsuspend users on their server.
As a result, moderation tooling such as Draupnir and Meowlnir have to implement
implementation-specific calls, which is not sustainable as more implementations other than Synapse
integrate this feature. This is also exacerbated by externally managed accounts that may be
provided by an entirely separate service.

This MSC will outline new endpoints that will allow server administrators to
suspend, unsuspend, lock, and unlock given users.

## Proposal

Note: what defines a "server administrator" is left up to the implementation,
most already have their own systems for defining administrators which rarely has a reason to be
exposed outside of their respective management interfaces.

> Taking advantage of the 10.22 section of the client-to-server specification,

Reviewer 7: ```suggestion
Taking advantage of the [10.22 section (Server Administration)](https://spec.matrix.org/v1.15/client-server-api/#server-administration) of the client-to-server specification,
```

Name and link to the section in the current spec, so that we know what it's talking about when the sections change numbers.

> four new endpoints are introduced:

Reviewer 2: A more complete user info endpoint would likely be useful, but it's out of scope for this MSC as you'd probably want to add a lot more information rather than just the new two booleans. It was discussed in this thread: https://github.com/matrix-org/matrix-spec-proposals/pull/4323#discussion_r2257823140

should be added to the alternatives section though

> four new endpoints are introduced:

Reviewer 4: I'd be happy to have it just listed in alternatives (or potential issues), yep!

- `GET /_matrix/client/v1/admin/suspend/{userId}`
- `PUT /_matrix/client/v1/admin/suspend/{userId}`
- `GET /_matrix/client/v1/admin/lock/{userId}`

> - `PUT /_matrix/client/v1/admin/lock/{userId}`

Reviewer 2: As per the checklist, authentication requirements (obviously), rate-limiting (none?) and guest access (no) should be specified explicitly.

These new endpoints complement the already existing
`/_matrix/client/v3/admin/whois/{userId}` endpoint, however, are restricted to only permitting
execution on local users.

The requests and responses are outlined as below. Remember that custom properties may be

> provided, given that they use proper namespaces.

Reviewer 8: I'm not quite sure what this means. Possibly you mean, "provided that they use"? (ie. custom properties can be included if obey the namespacing)

Reviewer: Reworded this, let me know if it's clearer now

The response body of both the `GET` and `PUT` endpoints, as well as the request body of the
`PUT` endpoint should be:

**Suspend**:

```json
{
  "$schema": "https://json-schema.org/draft-07/schema",
  "type": "object",
  "properties": {
    "suspended": {"type": "boolean"}
  },
  "required": ["suspended"]
}
```

For example:

```json
{"suspended": true}
```

**Lock**:

```json
{
  "$schema": "https://json-schema.org/draft-07/schema",
  "type": "object",
  "properties": {
    "locked": {"type": "boolean"}
  },
  "required": ["locked"]
}
```

For example:

```json
{"locked": false}
```

---

> Sending a `GET` to the respective endpoints should return `404` with `M_NOT_FOUND` if the

Reviewer 2: I think the usual way is to not use any RFC 2119 keywords, but rather just state it as a matter of fact

```suggestion
Sending a `GET` to the respective endpoints returns `404` with `M_NOT_FOUND` if the
```

(also need to fix the other similar "should"s)

requested user ID is not found. `400` and `M_INVALID_PARAM` should be returned if the specified
user ID is valid, but not local to the server. `403` `M_FORBIDDEN` should be returned if
the user is not a server administrator. To prevent enumeration, credentials should be checked

> before validity.

Reviewer 7: Are there any restrictions on who can be suspended/locked?  e.g. can an administrator suspend another administrator?  Or is it up to the implementation?  If someone tries to suspend someone they aren't allowed to, what error code gets returned?  (Presumably `M_FORBIDDEN`)

Reviewer: Hopefully all of this was answered in the latest iteration.

If a user is deactivated, `M_NOT_FOUND` should also be returned, as deactivated users are not
suspended nor locked, and typically are to be permanently treated as gone.

The server should advertise support for these endpoints by including the following new capability:

```json5
{
  "capabilities": {
    "account_moderation": {
      "suspend": true, // or false
      "lock": true // or false
    }
  }
}

> ```

Reviewer 8: Is this necessary? It would be unusual for endpoint I think. This suggests the endpoints are optional which I don't think is the intention? (and I don't think they should be). Also, trying and getting an error would generally be the way to detect whether the endpoint is implemented or not.

Reviewer 8: Hmm, using this for user permissions feels like a bit of a hack and maybe highlights the problem that server admin-ness is completely unspecified. If this is the case, could it be made explicit please? Furthermore, this MSC later talks about blocking endpoints to prevent admins from using them if desired, which somewhat undermines this.

Reviewer 7: For detecting support, could you use one of the `GET` endpoints on your own user ID?

Reviewer 2: > For detecting support, could you use one of the `GET` endpoints on your own user ID?

That sounds *way* more hacky than capabilities. Capabilities are already used to check permissions for things like profile/3pid/password changes, so I don't see why they couldn't be used for server-level moderation endpoints as well.

The server should only advertise these capabilities if the requesting user is a
server administrator, or otherwise has permission to suspend or lock users.

If the user is already in the state requested by a `PUT` request (e.g. a new
`PUT` request with `{"suspended":true}` is sent for a user that is already suspended),
the server should simply respond with `200 OK` as if a change was made. If metadata surrounding
the action was changed in the request (such as who locked the target), this should still be

> reflected.

Reviewer 8: I think this could do with being clearer: I'm not sure what it's saying it should reflect.

Servers may choose to retain implementation-specific versions of these endpoints, however should
still implement this uniform endpoint for compatibility.

## Potential issues

This MSC does not outline any additional fields for management, such as action reasons,
temporary actions, etc, which implementation-specific methods may currently have.
This is for the sake of brevity and simplicity, and may be expanded upon
in a future proposal. If implementations need further data, they may use custom properties,

> or retain their own endpoints and encourage operators to block these new ones.

Reviewer 8: Operators as in people who run the homeserver? This sounds like quite a strange thing to say in a spec and feels very much like a symptom of something more fundamentally wrong.

Reviewer 8: I'm still not quite sure what purpose this is trying to serve by being in the spec. If organisations want to do funny things blocking access to various endpoints, that's their prerogative, I don't really see why the spec needs to make allowances for them. If we do, it implies we need to start thinking about what happens if they do, eg. will the HS still advertise the capabilities but then the endpoints won't work?

Reviewer: I could likely slim it down a bit or make it less prominent, but I feel that this is still important context to include in the proposal itself somewhere as it adds two new endpoints that can potentially be used to cause at best a headache, at worst a full homeserver takeover, and noting that blocking the endpoint entirely is a valid precaution.

Potential issues don't go into the spec itself, no, but the MSCs are linked in the spec updates and I believe that proposals are supposed to outline all considerations taken in their design.

> If we do, it implies we need to start thinking about what happens if they do, eg. will the HS still advertise the capabilities but then the endpoints won't work?

Honestly, allowing servers to disable the capability advertisement might not be a bad idea, but I don't want to go too far into implementation specifics - otherwise we'll end up with a proposal that tells servers they need to allow enabling/disabling specific endpoints, etc, and then the proposal will never see the light of day.

## Alternatives

None

## Security considerations

Server operators who may not want their administrators to be able to suspend/lock users may have
restricted access to implementation-specific endpoints, usually by preventing access at the reverse
proxy. Operators should keep this in mind and apply similar restrictions to these new endpoints
where appropriate.

> 

Reviewer 7: An additional security consideration: An attacker who is able to obtain an access token for an administrator will be able to suspend and lock other accounts.
